docs: revamp information about interpolation

Narretz opened the pull request:
- Move interpolation info from Directive guide into new interpolation guide
- Add information about boolean attributes to interpolation guide
- remove wroong examples from prefixed boolean attribute docs, link
  to interpolation guide instead
- mention additional examples for attributes that benefit from ngAttr

I felt this was a good idea because currently interpolation information has no real place in the guide. So these docs try to bring the available info together and add some other stuff that has come up recently (ngAttr, mixing expressions and interpolation)

Narretz committed:
docs: reorganize information about interpolation

- Move interpolation info from Directive guide into new interpolation guide
- Add information about boolean attributes to interpolation guide
- remove wroong examples from prefixed boolean attribute docs, link
to interpolation guide instead
- mention additional examples for attributes that benefit from ngAttr
- add docs for ngRequired directive

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -0,0 +1,125 @@
+
+For this reason, Angular provides special `ng`-prefixed directives for the following boolean attributes:
+{@link ngDisabled `disabled`}, [@link ngSelected `selected`}, {@link ngChecked `checked`},
+{@link ngReadonly `readOnly`} , and [@link ngOpen `open`}.
```
`ngMultiple` and of course `ngRequired` :)

Narretz replied:
ngMultiple doesn't exist actually. At least not in attrs.js (also no docs). I think this is because binding to multiple in select is not supported, and in inputs it's not yet supported.

+1 for ngRequired

gkalpak replied:
Right. It is a `BOOLEAN_ATTR`, but I missed that we explicitly skip `multiple` in 'attrs.js'. So no `ngMultiple`.

gkalpak commented on docs/content/guide/interpolation.ngdoc:
````diff
@@ -0,0 +1,125 @@
+to true when the expression evaluates to truthy.
+
+ ```html
+  <div ng-init="isDisabled = false">
````
Woohoo, an anti-pattern :)

How about: 

```html
<div>
  Disabled: <input type="checkbox" ng-model="isDisabled" />
  <button ng-disabled="isDisabled">Click me</button>
</div>
```

gkalpak replied:
OK, it was copied over - but I would still change it.

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -0,0 +1,125 @@
+which will update the corresponding text nodes or attribute values as part of the
+normal {@link ng.$rootScope.Scope#$digest digest} cycle.
+
+### Binding to boolean attributes
```
This section is very useful for sure, but I don't think it is related to interpolation.
If there isn't a better place to put it though, it's much better here than nowhere :)

Narretz replied:
Yes, it's not strictly related. But one of the uses of interpolation is setting dynamic attribute values. And in that way it's relevant I'd say.

gkalpak replied:
Works for me (for lack of a better suggestion :smiley:).

gkalpak commented on docs/content/guide/interpolation.ngdoc:
````diff
@@ -0,0 +1,125 @@
+
+```html
+<svg>
+  <circle ng-attr-cx="cx"></circle>
````
Shouldn't it be `ng-attr-cx="{{ cx }}"` ?

Narretz replied:
Yes. I thought ngAttr works like the other ng- prefixed directives, but it's a separate mechanism.

gkalpak commented on docs/content/guide/interpolation.ngdoc:
````diff
@@ -0,0 +1,125 @@
+For example, to bind to `viewBox`, we can write:
+
+```html
+<svg ng-attr-view_box="viewBox">
````
_viewBox_  -->  _{{ viewBox }}_

gkalpak commented on docs/content/guide/interpolation.ngdoc:
````diff
@@ -0,0 +1,125 @@
+</svg>
+```
+
+The following attributes are also known to cause problems with normal bindings:
````
Could this be more clear ? What do you mean _"cause problems with normal bindings"_.

I guess it means one can't use interpolation directly on the attribute value and should use `ngAttr` instead, but could we be more explicit ?

Narretz replied:
There are issues for these two, maybe I can just link them?

gkalpak replied:
Probably works for me.

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -0,0 +1,125 @@
+- **placeholder** in `<textarea>` in Internet Explorer 10/11
+
+
+### Embedding interpolations inside expression
```
Not sure if it was intentional, but `expressions` sounds better to be.

Narretz replied:
+1

gkalpak commented on docs/content/guide/interpolation.ngdoc:
````diff
@@ -0,0 +1,125 @@
+<div ng-messages="items[{{index}}]"></div>
+```
+
+You should instead delegate the computation of complex expressions to the scope, like this:
````
Like what ?

Narretz replied:
Oops!

gkalpak commented on docs/content/guide/interpolation.ngdoc:
```diff
@@ -0,0 +1,125 @@
+### Embedding interpolations inside expression
+
+Angular directives take either expressions or interpolation markup with embedded expressions. So the
+following example which embeds interpolation inside an expression is a bad practice:
```
I would take it a step further and say that it might not work as expected (and even if it works now, it might break in the future).

gkalpak replied:
Sorry, I just saw it is mentioned below.

gkalpak commented on src/ng/compile.js:
```diff
@@ -1228,7 +1228,8 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
        * @param {function(interpolatedValue)} fn Function that will be called whenever
                 the interpolated value of the attribute changes.
-       *        See the {@link guide/directive#text-and-attribute-bindings Directives} guide for more info.
+       *        See the {@link guide/interpolation#how-text-and-attribute-bindings-work Directives}
```
_Directives_  -->  _Interpolation_ (?)

gkalpak replied:
You could also put "guide" inside the curly braces for consistency with other occurrences.

Narretz replied:
+1

gkalpak commented on src/ng/directive/attrs.js:
```diff
@@ -248,13 +229,12 @@
- * This complementary directive is not removed by the browser and so provides
- * a permanent reliable place to store the binding information.
+ *
+ * Sets the `readOnly` attribute on the element, if the expression inside `ngReadOnly` is truthy.
```
_ngReadOnly_  -->  _ngReadonly_

gkalpak commented:
Nice change !

A few suggestions (not directly related with this PR):

1. Add documentation for `ngRequired` and `ngMultiple` as standalone directives.
 2. Maybe remove the `ng-` prefixed versions of boolean attributes from the lists of attributes supported by the `input`/`select`/`textarea` directives. It's not like the controls make anything special to support the `ng`-prefixed versions.

Narretz commented:
Yes, the docs for ngRequired (and also ngMinlength and ngMaxlength) are overdue. They weren't always standalone directives, so that's why they are missing.

Narretz commented:
I've updated the docs and addressed your comments, and also added docs for ngRequired. PTAL. Removing the ng- prefixed directives from hte inputs - I thought about it, but to be consistent we would need to remove the unprefixed attributes too, because they invoke the same directive.

Narretz commented:
Landed as https://github.com/angular/angular.js/commit/8863836cd4b82c8f0f9d72de4323aace28402c38
